src/rmem.c

```diff
@@ -0,0 +1,470 @@
+        // make sure the pointer data is valid.
+        if( (ptrcmp)mem_node < (ptrcmp)mempool->stack.base || ((ptrcmp)mem_node - (ptrcmp)mempool->stack.mem) > mempool->stack.size || mem_node->size==0UL || mem_node->size > mempool->stack.size )
+            return;
+        // if the mem_node is right at the stack base ptr, then add it to the stack.
+        else if( (ptrcmp)mem_node == (ptrcmp)mempool->stack.base ) {
+            mempool->stack.base += mem_node->size;
+        }
+        // otherwise, we add it to the free list.
+        // We also check if the freelist already has the pointer so we can prevent double frees.
+        else if( mempool->freeList.len==0UL || ((ptrcmp)mempool->freeList.head >= (ptrcmp)mempool->stack.mem && (ptrcmp)mempool->freeList.head - (ptrcmp)mempool->stack.mem < mempool->stack.size) ) {
+            for( struct MemNode *n = mempool->freeList.head; n != NULL; n = n->next )
+                if( n==mem_node )
+                    return;
+            
+            // this code inserts at head.
+            /*
+            ( mempool->freeList.head==NULL )? (mempool->freeList.tail = mem_node) : (mempool->freeList.head->prev = mem_node);
+            mem_node->next = mempool->freeList.head;
+            mempool->freeList.head = mem_node;
+            mempool->freeList.len++;
+            */
+            
+            // this code insertion sorts where largest size is first.
+            if( mempool->freeList.head==NULL ) {
+                mempool->freeList.head = mempool->freeList.tail = mem_node;
+                mempool->freeList.len++;
+            } else if( mempool->freeList.head->size <= mem_node->size ) {
+                mem_node->next = mempool->freeList.head;
+                mem_node->next->prev = mem_node;
+                mempool->freeList.head = mem_node;
+                mempool->freeList.len++;
+            } else if( mempool->freeList.tail->size > mem_node->size ) {
+                mem_node->prev = mempool->freeList.tail;
+                mempool->freeList.tail->next = mem_node;
+                mempool->freeList.tail = mem_node;
+                mempool->freeList.len++;
+            } else {
+                struct MemNode *n = mempool->freeList.head;
+                while( n->next != NULL && n->next->size > mem_node->size )
+                    n = n->next;
+                
+                mem_node->next = n->next;
+                if( n->next != NULL )
+                    mem_node->next->prev = mem_node;
+                
+                n->next = mem_node;
+                mem_node->prev = n;
+                mempool->freeList.len++;
+            }
+            
+            if( mempool->freeList.autoDefrag && mempool->freeList.maxNodes != 0UL && mempool->freeList.len > mempool->freeList.maxNodes )
+                MemPool_DeFrag(mempool);
+        }
+    }
+}
+
+void MemPool_CleanUp(struct MemPool *const __restrict mempool, void *ptrref)
```

Comment:
* We presume C99, thus use restrict. If MSVC doesn't like it, add some #ifdefery on-top of the file that defines restrict to __restrict if necessary
* use double indirect pointer for the second argument, to make misuse harder?

Reply:
Problem with double indirect pointer is that you constantly need to cast the pointer type to `void**` to where that gets as verbose as `reinterpret_cast<void**>()`

Reply:
Ye, but on the other hand, a plain `void *` parameter is easily passed the pointer, not a ref. Both are suboptimal, but I'd argue better have the user be aware while casting, instead of erroneously accepting wrong pointers.